lib/helperFunctions.js

```diff
@@ -1,7 +1,93 @@
+import axios from "axios";
+import { aboutMe, itemsToFetch } from "../src/constants";
+
 export const scrollToSection = (id) => {
   const element = document.getElementById(id);
   const yOffset = -70; // Adjust this value based on your header height
   const y = element.getBoundingClientRect().top + window.scrollY + yOffset;
 
   window.scrollTo({ top: y, behavior: "smooth" });
 };
+
+const parseOriginFromUrl = (url) => {
+  /**
+   * splits https://api.github.com/repos/org-name/repo-name/issues/25
+   * into ["https://", "api.github.com", "repos", "org-name", "repo-name", "issues", "25"]
+   */
+  const parts = url.split(/https:\/\/|\//gm);
+  const organization = parts[3];
+  const repo = parts[4];
+  /**
+   * accessing a github profile or organization and adding a .png
+   * at the end of the URL will return their logo/profile picture
+   */
+  const logoUrl = `https://github.com/${organization}.png`;
+
+  return {
+    organization,
+    repo,
+    logoUrl,
+  };
+};
+
+export async function fetchContributions() {
+  try {
+    const response = await axios.get("https://api.github.com/search/issues", {
+      params: {
+        q: `author:${aboutMe.githubUsername}`,
+        sort: "created",
+        order: "desc",
+        per_page: itemsToFetch,
+      },
+      headers: {
+        Authorization: `token ${import.meta.env.VITE_GH_TOKEN}`,
+      },
+    });
+
+    const items = response.data.items;
+
+    const filteredContributions = await Promise.all(
+      items.map(async (item) => {
+        let linesAdded = 0;
+        let linesDeleted = 0;
+        if (item.pull_request) {
+          const detailsResponse = await axios.get(item.pull_request.url, {
+            headers: {
+              Authorization: `token ${import.meta.env.VITE_GH_TOKEN}`,
+            },
+          });
+          linesAdded = detailsResponse.data.additions || 0;
+          linesDeleted = detailsResponse.data.deletions || 0;
+        }
+
+        const { organization, repo, logoUrl } = parseOriginFromUrl(item.url);
+
+        if (organization === aboutMe.githubUsername) {
+          return null;
+        }
+
+        return {
+          id: item.id,
+          organisation: organization,
+          logo: logoUrl,
+          repo: repo,
+          type: item.pull_request ? "Pull Requests" : "Issues",
+          status: item.state,
+          title: item.title,
+          link: item.html_url,
+          number: `#${item.number}`,
+          date: new Date(item.created_at).toLocaleDateString(),
+          linesAdded,
+          linesDeleted,
+        };
+      })
+    );
+
+    return filteredContributions
+      .filter((contribution) => contribution != null)
+      .slice(0, 12);
```

Comment:
This should also be `itemsToFetch`?

Reply:
No, 
Let's say I fetch 12 items, then out of those 5 of them were from your account, after filtering those out we'll be left with 7. Showing 7 contributions looked a empty on portfolio. So I always fetch extra items(20), filter them and show 12 of them to make sure the section doesn't look empty.

Reply:
@A-4-Atom Ok so this is not a neat approach and it leads to the issue I was guessing.

> The current implementation would fetch even the issues and PRs I make in my own repositories. For instance, I might be working on some project of mine, leading to multiple PRs and issues. This would flood this section with those made recently.

As I explained above, the 20 that you fetch are all from my own projects. So when it filters it out, its left with nothing. I had to push the limit to 100 to be able to see contributions that I wanted to see. 

I think in this case, the best approach would be to specify the repositories. It gets the best of both worlds. Plus, the user would have control over what they want to display. More recent activity might not mean the best of what they would have done in the past. One might have contributed to like 10 repos out of which they would want to showcase the contributions from only the 2nd, 5th and 7th one.

Also, lets add one retry as well in case of a failure.

Reply:
Hi, I've created an array of strings in **/constants/index** file _includedRepos_.
It is used when fetching from github API, user can specifically add whichever repos they want pr and issues from.
Also I've made so that if an error occurs, it retries one more time.
@mittal-parth

Reply:
Thanks @A-4-Atom! Checked it, better.

Although I noticed one thing which I somehow overlooked before.

We are making a request for every pr to get its detail? That's quite non-performant.
The perfect solution here is to use the GraphQL API which GitHub already exposes. We can get the relevant information (and only the information we need) in a single request. That's the power of GraphQL wherein it prevents under fetching and over fetching.

Lemme know if you want me to push the changes, I did a little tinkering already to check its feasibility.

Reply:
I had to make separate API Calls for pr's to get number of lines added/deleted, as it wasn't included in he original fetch.
What if we fetch the pr and issues separately so in total only 2 api calls will be made.
@mittal-parth

Reply:
> I had to make separate API Calls for pr's to get number of lines added/deleted, as it wasn't included in he original fetch.

Yup I understood this.

> What if we fetch the pr and issues separately so in total only 2 api calls will be made.

How? We still have to make individual requests for the PR to get the lines info?

Reply:
Also, I am thinking to get away with issues. Just PRs would be nicer imo.

And then the filtering would be based on the repo name.

Reply:
Let me try if it works, then i'll get back to you.

Reply:
Lets be clear on what you are trying? @A-4-Atom

Reply:
I was thinking if we just get the pr instead of pr+issues it will return the no. of lines added and deleted.
I just tried it did not work.

Reply:
Yeah that wont work. The response wont change.

Lets stick to GraphQL. Lemme know if you want to/ have the bandwidth to do it. Otherwise I can push the changes as well.

Reply:
@mittal-parth can we fetch all of the required details in a single fetch with graphQl? I've never used graphQl.

Reply:
yes we can, thats what I am trying to outline here :)

Reply:
GraphQL has only one endpoint.

Its more like a query that you send to the server with exactly what you need. 

And it returns that.

Reply:
Okay, I'll take a look at the graphQL's documentation and update the code accordingly.

Reply:
Cool, thanks!

Also make sure that we are now filtering based on the repos and not 'Issues and PRs'. We can get rid of issues altogether.

Reply:
Just to make sure, You want to fetch just the pr's from selected repos, and from there users can filter by clicking on the repo name (like how it's currently on your website).